test/test_nestedtensor.py

```diff
@@ -0,0 +1,210 @@
+                                     getattr(torch, func)(b, c)])
+            assert (a3 == getattr(torch, func)(a1, a2)).all()
+            assert not (a3 == a1).any()
+            assert not (a3 == a2).any()
+            assert (a3 == getattr(a1, func)(a2)).all()
+            assert not (a3 == a1).any()
+            assert not (a3 == a2).any()
+            assert (a3 == getattr(a1, func + "_")(a2)).all()
+            assert (a3 == a1).all()
+
+    def test_detach(self):
+        data = [self.gen_float_tensor(1, (10, 10)),
+                self.gen_float_tensor(2, (10, 10)),
+                self.gen_float_tensor(3, (10, 10))]
+        ones_data = [torch.ones(10, 10),
+                     torch.ones(10, 10),
+                     torch.ones(10, 10)]
+        # We don't support scalar arguments yet (broadcasting)
+        # This will be part of NestedTensor 0.0.3
+        twos_data = [torch.ones(10, 10) * 2,
+                     torch.ones(10, 10) * 2,
+                     torch.ones(10, 10) * 2]
+        fours_data = [torch.ones(10, 10) * 4,
+                      torch.ones(10, 10) * 4,
+                      torch.ones(10, 10) * 4]
+        ones = torch.nestedtensor(ones_data).to(torch.float)
+        twos = torch.nestedtensor(twos_data).to(torch.float)
+        fours = torch.nestedtensor(fours_data).to(torch.float)
+        x = torch.nestedtensor(data, requires_grad=True)
+        y = x + twos
+        y = y.detach()
+        z = y * fours + twos
+        self.assertFalse(y.requires_grad)
+        self.assertFalse(z.requires_grad)
+
+        x = torch.nestedtensor(data, requires_grad=True)
+        y = x * twos
+        y = y.detach()
+        self.assertFalse(y.requires_grad)
+        self.assertIsNone(y.grad_fn)
+        z = x + y
+        z.sum().backward()
+
+        # This is an incorrect gradient, but we assume that's what the user
+        # wanted. detach() is an advanced option.
+        self.assertTrue((x.grad.data == ones).all())
+
+        # in-place detach
+        x = torch.nestedtensor(data, requires_grad=True)
+        y = torch.nestedtensor(data, requires_grad=True)
+        a = x * twos
+        (y + a).sum().backward(retain_graph=True)
+        a.detach_()
+        self.assertFalse(a.requires_grad)
+        (y + a).sum().backward()  # this won't backprop to x
+        self.assertTrue((x.grad.data == ones * twos).all())
+        self.assertTrue((y.grad.data == ones * twos).all())
+
+        # TODO: view semantics will be defined by NestedTensor 0.0.3 or 0.0.4
+        # in-place deatch on a view raises an exception
+        # view = x.narrow(0, 1, 4)
+        # self.assertRaisesRegex(RuntimeError, 'view', lambda: view.detach_())
+
+    def test_detach_base(self):
+        "detaching base does not detach view"
+        x = torch.nestedtensor([torch.randn(10, 10)], requires_grad=True)
+        x.detach_()
+        self.assertFalse(x.requires_grad)
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
You might want to add this to the list of all tests eventually: https://github.com/pytorch/pytorch/blob/master/test/run_test.py

Reply:
Not just eventually, in this PR :>

Reply:
I really want to prevent having NestedTensors show up by default. This is another point of failure for the sake of completeness for an experimental feature.

Reply:
I don't see how properly testing code counts as having it "show up by default"? Also, having tests that run on CI is a way of protecting yourself as well: if someone makes a change that breaks NestedTensor, they have no way of finding out if CI doesn't find it. When XLA lived in this universe, responsibility fell solely on XLA developers to fix breakages caused by other developers. (And I don't think we're likely to be breaking nested tensor, not at least in its current formulation.)

Reply:
Resolved offline: We can run this test in a separate process to make sure the monkeypatching doesn't affect other tests.